src/Features/Core/Portable/DisposeAnalysis/DisposableFieldsShouldBeDisposedDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,266 @@
+                // already seen dispose invocations, i.e. currentValue = true, continue to be marked as disposed.
+                _fieldDisposeValueMap.AddOrUpdate(field,
+                    addValue: disposed,
+                    updateValueFactory: (f, currentValue) => currentValue || disposed);
+            }
+
+            private void OnSymbolEnd(SymbolAnalysisContext symbolEndContext)
+            {
+                foreach (var kvp in _fieldDisposeValueMap)
+                {
+                    IFieldSymbol field = kvp.Key;
+                    bool disposed = kvp.Value;
+                    if (!disposed)
+                    {
+                        // Disposable field '{0}' is never disposed
+                        var diagnostic = Diagnostic.Create(s_disposableFieldsShouldBeDisposedRule, field.Locations[0], field.Name);
+                        symbolEndContext.ReportDiagnostic(diagnostic);
+                    }
+                }
+            }
+
+            private void OnOperationBlockStart(OperationBlockStartAnalysisContext operationBlockStartContext)
+            {
+                switch (operationBlockStartContext.OwningSymbol)
+                {
+                    case IFieldSymbol _:
+                        // Field initializer.
+                        if (operationBlockStartContext.OperationBlocks.Length == 1 &&
+                            operationBlockStartContext.OperationBlocks[0] is IFieldInitializerOperation fieldInitializer)
+                        {
+                            foreach (var field in fieldInitializer.InitializedFields)
+                            {
+                                if (!field.IsStatic && _disposableFields.Contains(field))
+                                {
+                                    // Instance field initialized with a disposable object is considered a candidate.
+                                    AddOrUpdateFieldDisposedValue(field, disposed: false);
+                                }
+                            }
+                        }
+
+                        break;
+
+                    case IMethodSymbol containingMethod:
+                        // Method body.
+                        OnMethodOperationBlockStart(operationBlockStartContext, containingMethod);
+                        break;
+                }
+            }
+
+            private void OnMethodOperationBlockStart(OperationBlockStartAnalysisContext operationBlockStartContext, IMethodSymbol containingMethod)
+            {
+                PointsToAnalysisResult lazyPointsToAnalysisResult = null;
+
+                // If we have any potential disposable object creation descendant within the operation blocks,
+                // register an operation action to analyze field references where field might be assigned a disposable object.
+                if (_disposeAnalysisHelper.HasAnyDisposableCreationDescendant(operationBlockStartContext.OperationBlocks, containingMethod))
```

Comment:
I guess this take cares of things like lambda/local function/anonymous method?

Reply:
Yes, it walks the whole operation tree looking for specific operation kinds such as invocations that return IDisposable, object creations of disposable types, etc. See [here](https://github.com/dotnet/roslyn-analyzers/blob/0bb8197ff0551722a507027d8e3c7a97a5e62a9e/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysisHelper.cs#L131)